File: ios/Sources/ACRendering/Events/CardLifecycleEvent.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import Foundation
import CoreGraphics
import ACCore

/// Lifecycle events the host may observe via `onCardLifecycle`.
public enum CardLifecycleEvent {
    /// Card body has been rendered (layout complete, images may still be loading)
    case rendered

    /// Card's intrinsic content size changed (e.g., ShowCard expanded, image loaded)
    case sizeChanged(CGSize)

    /// An input value changed
    case inputChanged(id: String, value: Any)

    /// Parse failed (useful for the json: convenience init)
    case parseFailed(ParseError)

    /// Performance report — fires after the card is fully loaded
    case performanceReport(CardPerformanceMetrics)
}

File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/CardLifecycleEvent.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.adaptivecards.core

/**
 * Lifecycle events the host may observe via `onLifecycle`.
 */
sealed interface CardLifecycleEvent {
    /** Card body has been rendered (layout complete, images may still be loading) */
    data object Rendered : CardLifecycleEvent

    /** Card's intrinsic content size changed */
    data class SizeChanged(val width: Float, val height: Float) : CardLifecycleEvent

    /** An input value changed */
    data class InputChanged(val id: String, val value: Any) : CardLifecycleEvent

    /** Parse failed */
    data class ParseFailed(val error: ParseError) : CardLifecycleEvent

    /** Performance report — fires after the card is fully loaded */
    data class PerformanceReport(val metrics: CardPerformanceMetrics) : CardLifecycleEvent
}
